plugins/modem-manager/fu-qmi-pdc-updater.c

```diff
@@ -0,0 +1,432 @@
+		return;
+	}
+
+	if (!qmi_indication_pdc_load_config_output_get_remaining_size (output, &remaining_size, &ctx->error)) {
+		g_prefix_error (&ctx->error, "couldn't load remaining size: ");
+		g_main_loop_quit (ctx->mainloop);
+		return;
+	}
+
+	if (remaining_size == 0) {
+		g_debug ("finished loading mcfg");
+		g_main_loop_quit (ctx->mainloop);
+		return;
+	}
+
+	g_debug ("loading next chunk (%u bytes remaining)", remaining_size);
+	fu_qmi_pdc_updater_load_config (ctx);
+}
+
+static void
+fu_qmi_pdc_updater_load_config_ready (GObject *qmi_client, GAsyncResult *res, gpointer user_data)
+{
+	WriteContext *ctx = (WriteContext *) user_data;
+	g_autoptr(QmiMessagePdcLoadConfigOutput) output = NULL;
+
+	output = qmi_client_pdc_load_config_finish (QMI_CLIENT_PDC (qmi_client), res, &ctx->error);
+	if (output == NULL) {
+		g_main_loop_quit (ctx->mainloop);
+		return;
+	}
+
+	if (!qmi_message_pdc_load_config_output_get_result (output, &ctx->error)) {
+		g_main_loop_quit (ctx->mainloop);
+		return;
+	}
+
+	/* after receiving the response to our request, we now expect an indication
+	 * with the actual result of the operation */
+	g_assert (ctx->indication_id == 0);
+	ctx->indication_id = g_signal_connect (ctx->qmi_client, "load-config",
+					       G_CALLBACK (fu_qmi_pdc_updater_load_config_indication), ctx);
+
+	/* don't wait forever */
+	g_assert (ctx->timeout_id == 0);
+	ctx->timeout_id = g_timeout_add_seconds (5, fu_qmi_pdc_updater_load_config_timeout, ctx);
+}
+
+static void
+fu_qmi_pdc_updater_load_config (WriteContext *ctx)
+{
+	g_autoptr(QmiMessagePdcLoadConfigInput) input = NULL;
+	g_autoptr(GArray) chunk = NULL;
+	gsize full_size;
+	gsize chunk_size;
+
+	input = qmi_message_pdc_load_config_input_new ();
+	qmi_message_pdc_load_config_input_set_token (input, ctx->token++, NULL);
+
+	full_size = g_bytes_get_size (ctx->blob);
+	chunk_size = (((ctx->offset + QMI_LOAD_CHUNK_SIZE) > full_size) ?
+		      (full_size - ctx->offset) :
+		      QMI_LOAD_CHUNK_SIZE);
+
+	chunk = g_array_sized_new (FALSE, FALSE, sizeof (guint8), chunk_size);
+	g_array_set_size (chunk, chunk_size);
+	memcpy (chunk->data, (const guint8 *)g_bytes_get_data (ctx->blob, NULL) + ctx->offset, chunk_size);
+
+	qmi_message_pdc_load_config_input_set_config_chunk (input,
+							    QMI_PDC_CONFIGURATION_TYPE_SOFTWARE,
+							    ctx->digest,
+							    full_size,
+							    chunk,
+							    NULL);
+	ctx->offset += chunk_size;
+
+	qmi_client_pdc_load_config (ctx->qmi_client, input, 10, NULL, fu_qmi_pdc_updater_load_config_ready, ctx);
+}
+
+gboolean
+fu_qmi_pdc_updater_write (FuQmiPdcUpdater *self, const gchar *filename, GBytes *blob, GError **error)
+{
+	g_autoptr(GMainLoop) mainloop = g_main_loop_new (NULL, FALSE);
+	g_autoptr(GChecksum) checksum = NULL;
+	g_autoptr(GArray) digest = NULL;
+	gsize file_size;
+	gsize hash_size;
+	WriteContext ctx = {
+		.mainloop = mainloop,
+		.qmi_client = self->qmi_client,
+		.blob = blob,
+	};
+
+	/* get checksum, to be used as unique id */
```

Comment:
maybe split this out into a helper function?

Reply:
ok